python/gtsam/examples/IMUKittiExampleGPS.py

```diff
@@ -0,0 +1,332 @@
+            gps_measurement = GpsMeasurement(time, np.asarray([x, y, z]))
+            gps_measurements.append(gps_measurement)
+
+    return kitti_calibration, imu_measurements, gps_measurements
+
+
+def getImuParams(kitti_calibration: KittiCalibration):
+    """Get the IMU parameters from the KITTI calibration data."""
+    GRAVITY = 9.8
+    w_coriolis = np.zeros(3)
+
+    # Set IMU preintegration parameters
+    measured_acc_cov = np.eye(3) * np.power(
+        kitti_calibration.accelerometer_sigma, 2)
+    measured_omega_cov = np.eye(3) * np.power(
+        kitti_calibration.gyroscope_sigma, 2)
+    # error committed in integrating position from velocities
+    integration_error_cov = np.eye(3) * np.power(
+        kitti_calibration.integration_sigma, 2)
+
+    imu_params = gtsam.PreintegrationParams.MakeSharedU(GRAVITY)
+    # acc white noise in continuous
+    imu_params.setAccelerometerCovariance(measured_acc_cov)
+    # integration uncertainty continuous
+    imu_params.setIntegrationCovariance(integration_error_cov)
+    # gyro white noise in continuous
+    imu_params.setGyroscopeCovariance(measured_omega_cov)
+    imu_params.setOmegaCoriolis(w_coriolis)
+
+    return imu_params
+
+
+def save_results(isam: gtsam.ISAM2, output_filename: str, first_gps_pose: int,
+                 gps_measurements: List[GpsMeasurement]):
+    """Write the results from `isam` to `output_filename`."""
+    # Save results to file
+    print("Writing results to file...")
+    with open(output_filename, 'w', encoding='UTF-8') as fp_out:
+        fp_out.write(
+            "#time(s),x(m),y(m),z(m),qx,qy,qz,qw,gt_x(m),gt_y(m),gt_z(m)\n")
+
+        result = isam.calculateEstimate()
+        for i in range(first_gps_pose, len(gps_measurements)):
+            pose_key = X(i)
+            vel_key = V(i)
+            bias_key = B(i)
+
+            pose = result.atPose3(pose_key)
+            velocity = result.atVector(vel_key)
+            bias = result.atConstantBias(bias_key)
+
+            pose_quat = pose.rotation().toQuaternion()
+            gps = gps_measurements[i].position
+
+            print(f"State at #{i}")
+            print(f"Pose:\n{pose}")
+            print(f"Velocity:\n{velocity}")
+            print(f"Bias:\n{bias}")
+
+            fp_out.write("{},{},{},{},{},{},{},{},{},{},{}\n".format(
+                gps_measurements[i].time, pose.x(), pose.y(), pose.z(),
+                pose_quat.x(), pose_quat.y(), pose_quat.z(), pose_quat.w(),
+                gps[0], gps[1], gps[2]))
+
+
+def parse_args():
+    """Parse command line arguments."""
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--output_filename",
+                        default="IMUKittiExampleGPSResults.csv")
+    return parser.parse_args()
+
+
+def main():
+    """Main runner."""
+    args = parse_args()
+    kitti_calibration, imu_measurements, gps_measurements = loadKittiData()
+
+    if not kitti_calibration.bodyTimu.equals(Pose3(), 1e-8):
```

Comment:
the code looks good. I think the main method is about 161 lines though, in length. So I would vote to modularize it a bit before checking it in. However, I do understand that the original C++ code main function was also very long, so not your fault in any way